Remove atoms parameter in Cantherm and allow user-defined atomic energies

cgrambow opened the pull request:
When parsing quantum chemistry output in a Cantherm calculation, there is no reason for the user to specify the types and numbers of atoms, because the information can be obtained from the output files. In fact, mistyping atoms will lead to incorrect thermochemistry and rates, which is avoided in this PR.

This PR also increases the flexibility of Cantherm calculations by allowing the user to supply a dictionary of atomic energies, which means that we are no longer restricted by the hard-coded model chemistries in `rmgpy/cantherm/statmech.py`.

Lastly, another issue in #1315 and #742 is addressed by removing model chemistries that do not specify which basis set was used for the atomic energy calculations. This is a relatively drastic measure, but it should ensure that users do not accidentally end up with completely wrong results.

alongd commented on documentation/source/users/cantherm/input.rst:
```diff
@@ -128,12 +127,11 @@ Parameter               Required?                   Description
+to apply atomization energy corrections (AEC) and spin orbit corrections (SOC) for a given ``modelChemistry()``
+(see `Model Chemistry`_).
 
-Allowed atom symbols for the ``atoms`` parameter are
```
We may still want to leave some description in the documentation of the atoms we commonly have AEC for

alongd commented on rmgpy/cantherm/statmech.py:
```diff
@@ -318,6 +311,16 @@ def load(self):
+        # Infer atoms from geometry
+        atoms = {}
+        for atom_num in number:
+            symbol = atom_num_dict[atom_num]
```
We should capture exception when `atom_num` is not in `atom_num_dic`, and output an error message that this atom is currently not supported in Cantherm

cgrambow replied:
Done. Since we're now allowing the user to input atomic energies, I also added more enthalpies of formation of atoms so that we now support more elements.

alongd commented on rmgpy/cantherm/statmech.py:
```diff
@@ -318,6 +311,16 @@ def load(self):
+        atoms = {}
+        for atom_num in number:
+            symbol = atom_num_dict[atom_num]
+            atoms[symbol] = atoms.get(symbol, 0) + 1
```
excellent implementation!

alongd commented:
In the spirit of this PR, could you also add a tiny fix for annoying cases when the bond correction is entered in reverse, e.g., `O=C` instead of `C=O`?
I believe [this line](https://github.com/ReactionMechanismGenerator/RMG-Py/blob/master/rmgpy/cantherm/statmech.py#L721) should be changed to also check if `symbol[::-1]` is in `bondEnergies`.
Thanks!

cgrambow commented:
Should be good to go!

alongd commented on documentation/source/users/cantherm/input.rst:
```diff
@@ -13,12 +13,13 @@ Each section is made up of one or more function calls, where parameters are
 either single or double quotes.
 
-The following is a list of all the functions of a CanTherm input file for thermodynamics and high-pressure limit kinetics computations:
+The following is a list of all the elements of a CanTherm input file for thermodynamics and high-pressure limit kinetics computations:
```
I'm uncertain regarding this terminology change: "elements" sound like chemical elements. By the way, the table below still says `Function`, as well as other references in the documentation below. Since it's a python file, it's OK to call them functions. But we could consider "components" as well.

cgrambow replied:
The reason I changed it is because many of the options aren't functions, they're assignment statements. That was probably changed in the past, but the documentation never got updated. I thought I went through and checked for all occurrences of "function", but evidently I wasn't thorough enough. I agree that "elements" could be misleading. I'll give it a different name.

alongd commented on documentation/source/users/cantherm/input.rst:
```diff
@@ -33,16 +34,27 @@ Function                    Description
-    modelChemistry("CBS-QB3")
+    modelChemistry = "CBS-QB3"
+
+Alternatively, the atomic energies at the ``modelChemistry`` level of theory can be directly
```
We should tell the users that no bond corrections will be applied in this case, and perhaps allow them to add their own?

cgrambow replied:
Bond corrections would still be applied if `modelChemistry` matches one of the options in `applyEnergyCorrections`, although of course it is unlikely that someone would specify `atomEnergies` themselves if they are already available within Cantherm. I could add an option to add bond corrections, but deriving them is typically associated with a lot more work than calculating atomic energies, so I doubt that option would be used. Ideally, if new bond corrections become available, they should be added to Cantherm. What do you think?

alongd replied:
I think it's enough to mention in the documentation something like:
If a `modelChemistry` other than the ones specified in the table is given, then the user should add the respective `atomEnergies` to get meaningful results, and BAC would not be applied.

alongd commented on documentation/source/users/cantherm/input.rst:
```diff
@@ -131,18 +139,17 @@ Parameter               Required?                   Description
-The ``atom`` and ``bond`` parameters are used to apply atomization energy corrections (AEC), bond corrections (BC), and spin orbit corrections (SOC) for a given ``modelChemistry()`` (see `Model Chemistry`_).
-
-Allowed atom symbols for the ``atoms`` parameter are
-``'C'``, ``'N'``, ``'O'``, ``'S'``, ``'P'``, and ``'H'``. For example, for formaldehyde we would write::
```
We may still want to leave some description in the documentation of the atoms we commonly have AEC for

cgrambow replied:
Ah, I wasn't sure what you meant by this earlier, because the information about which AECs are available is still in the table, which I didn't delete. You mean we should have a description about which atomic energies are available for the implemented model chemistries, correct? I'll add that.

alongd replied:
Yes, something like what elements Canthem supports (from the atomHf dictionary you elaborated). But in second though, each modelChemistry supports slightly different elements... Perhaps we could list the common ones?
And more importantly, what do you think of outputting an Error is we don't have one of the elements within the `atomEnergies` dictionary for the selected modelChemistry (or within the user inputted atomEnergies values)?

cgrambow replied:
I don't think we should output that error because if the user is only interested in kinetics and doesn't care about the thermo, then it doesn't matter if atomic energies are available or not because they will cancel out between the products and reactants. Cantherm already prints a warning if that is the case, so at least the user would be aware of the issue.

alongd commented on rmgpy/cantherm/statmech.py:
```diff
@@ -318,6 +314,19 @@ def load(self):
+            try:
+                symbol = atom_num_dict[atom_num]
+            except KeyError:
+                raise Exception('Atom number {} is not yet supported.'.format(atom_num))
```
The error message needs a tiny refinement: "atom number" could also be inferred as the number of the atom in the xyz / Z matrix. Could you print the element instead?

cgrambow replied:
Yup, I'll change it

cgrambow commented:
Updated documentation, let me know what you think.

alongd commented on documentation/source/users/cantherm/input.rst:
```diff
@@ -33,61 +34,75 @@ Function                    Description
+for a given model chemistry.
+
+================================================ ===== ==== ==== ==================
+Model Chemistry                                  AEC   BC   SOC  Atom Types
```
I love this addition!
Could you just change "Atom Types" to something like "Supported elements"? "Atom Types" means something else in RMG.

cgrambow replied:
Sure thing! Too much RMG lingo...

cgrambow replied:
Done

cgrambow commented:
One thing I was wondering about: Cantherm prints a warning if an atom type/element is unknown, which will cause it to not subtract the atomic reference energy for that atom. In that case, I feel like the returned energy will not make any physical sense, i.e., it will not be useful for thermo, but should still be relevant for kinetics. If that happens, should we just print a warning that the energy corrections will not be applied and that the user should not trust the thermo values?

Cantherm also prints the equilibrium constant and reverse rates in comments in the output file. Those would not be affected, correct?

alongd commented:
Is there a reason for a user to only use partial atom energy corrections?
I think Cantherm should crush if AEC is used, but is partial.
So perhaps we need a flag for `useAtomEnergyCorrections` (like `useBondCorrections`), that defaults to True and allows Cantherm to be consciously used with no corrections if it is False?

cgrambow commented:
I made the changes. `useAtomCorrections` is `True` by default and Cantherm raises an error if elements are not found in the `atomEnergies` dictionary. I also added this to the documentation.

cgrambow commented:
Rebased

cgrambow commented:
I added the atomic energies for B3LYP/6-311+G(3df,2p). While doing so, I noticed that most of the model chemistry atomic energies in `rmgpy/cantherm/statmech.py` don't actually match the outputs of single-atom quantum calculations, so they probably were fitted in some way in the past, but that information has since been lost. We should try to find out where the values come from and start adding some documentation for posterity, but that can be done in future pull requests.

I also changed the code such that model chemistries are now case-insensitive. This means that future developers should make sure to add atomic energies and bond correction model chemistries in lower-case in `statmech.py`.
